File: src/Atlasd/Battlenet/Protocols/Game/Message.cs
using Atlasd.Battlenet.Protocols.Game.Messages;
using Atlasd.Battlenet.Exceptions;
using System;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game
{
    abstract class Message
    {
        public byte Id;
        public byte[] Buffer { get; protected set; }

        public static Message FromByteArray(byte[] buffer)
        {
            if (buffer[0] != 0xFF)
                throw new GameProtocolViolationException(null, "Invalid message header");

            byte id = buffer[1];
            UInt16 length = (UInt16)((buffer[3] << 8) + buffer[2]);
            byte[] body = new byte[length - 4];

            System.Buffer.BlockCopy(buffer, 4, body, 0, length - 4);

            return FromByteArray(id, body);
        }

        public static Message FromByteArray(byte id, byte[] buffer)
        {
            return ((MessageIds)id) switch
            {
                MessageIds.SID_NULL => new SID_NULL(buffer),
                MessageIds.SID_STOPADV => new SID_STOPADV(buffer),
                MessageIds.SID_CLIENTID => new SID_CLIENTID(buffer),
                MessageIds.SID_STARTVERSIONING => new SID_STARTVERSIONING(buffer),
                MessageIds.SID_REPORTVERSION => new SID_REPORTVERSION(buffer),
                MessageIds.SID_STARTADVEX => new SID_STARTADVEX(buffer),
                MessageIds.SID_GETADVLISTEX => new SID_GETADVLISTEX(buffer),
                MessageIds.SID_ENTERCHAT => new SID_ENTERCHAT(buffer),
                MessageIds.SID_GETCHANNELLIST => new SID_GETCHANNELLIST(buffer),
                MessageIds.SID_JOINCHANNEL => new SID_JOINCHANNEL(buffer),
                MessageIds.SID_CHATCOMMAND => new SID_CHATCOMMAND(buffer),
                MessageIds.SID_CHATEVENT => new SID_CHATEVENT(buffer),
                MessageIds.SID_LEAVECHAT => new SID_LEAVECHAT(buffer),
                MessageIds.SID_LOCALEINFO => new SID_LOCALEINFO(buffer),
                MessageIds.SID_FLOODDETECTED => new SID_FLOODDETECTED(buffer),
                MessageIds.SID_UDPPINGRESPONSE => new SID_UDPPINGRESPONSE(buffer),
                MessageIds.SID_CHECKAD => new SID_CHECKAD(buffer),
                MessageIds.SID_CLICKAD => new SID_CLICKAD(buffer),
                MessageIds.SID_READMEMORY => new SID_READMEMORY(buffer),
                MessageIds.SID_REGISTRY => new SID_REGISTRY(buffer),
                MessageIds.SID_MESSAGEBOX => new SID_MESSAGEBOX(buffer),
                MessageIds.SID_STARTADVEX2 => new SID_STARTADVEX2(buffer),
                MessageIds.SID_GAMEDATAADDRESS => new SID_GAMEDATAADDRESS(buffer),
                MessageIds.SID_STARTADVEX3 => new SID_STARTADVEX3(buffer),
                MessageIds.SID_LOGONCHALLENGEEX => new SID_LOGONCHALLENGEEX(buffer),
                MessageIds.SID_CLIENTID2 => new SID_CLIENTID2(buffer),
                MessageIds.SID_LEAVEGAME => new SID_LEAVEGAME(buffer),
                MessageIds.SID_DISPLAYAD => new SID_DISPLAYAD(buffer),
                MessageIds.SID_NOTIFYJOIN => new SID_NOTIFYJOIN(buffer),
                MessageIds.SID_PING => new SID_PING(buffer),
                MessageIds.SID_READUSERDATA => new SID_READUSERDATA(buffer),
                MessageIds.SID_WRITEUSERDATA => new SID_WRITEUSERDATA(buffer),
                MessageIds.SID_LOGONCHALLENGE => new SID_LOGONCHALLENGE(buffer),
                MessageIds.SID_LOGONRESPONSE => new SID_LOGONRESPONSE(buffer),
                MessageIds.SID_CREATEACCOUNT => new SID_CREATEACCOUNT(buffer),
                MessageIds.SID_SYSTEMINFO => new SID_SYSTEMINFO(buffer),
                MessageIds.SID_GAMERESULT => new SID_GAMERESULT(buffer),
                MessageIds.SID_GETICONDATA => new SID_GETICONDATA(buffer),
                MessageIds.SID_CDKEY => new SID_CDKEY(buffer),
                MessageIds.SID_CHANGEPASSWORD => new SID_CHANGEPASSWORD(buffer),
                MessageIds.SID_CHECKDATAFILE => new SID_CHECKDATAFILE(buffer),
                MessageIds.SID_GETFILETIME => new SID_GETFILETIME(buffer),
                MessageIds.SID_QUERYREALMS => new SID_QUERYREALMS(buffer),
                MessageIds.SID_CDKEY2 => new SID_CDKEY2(buffer),
                MessageIds.SID_LOGONRESPONSE2 => new SID_LOGONRESPONSE2(buffer),
                MessageIds.SID_CHECKDATAFILE2 => new SID_CHECKDATAFILE2(buffer),
                MessageIds.SID_CREATEACCOUNT2 => new SID_CREATEACCOUNT2(buffer),
                MessageIds.SID_LOGONREALMEX => new SID_LOGONREALMEX(buffer),
                MessageIds.SID_QUERYREALMS2 => new SID_QUERYREALMS2(buffer),
                MessageIds.SID_QUERYADURL => new SID_QUERYADURL(buffer),
                MessageIds.SID_WARCRAFTGENERAL => new SID_WARCRAFTGENERAL(buffer),
                MessageIds.SID_NEWS_INFO => new SID_NEWS_INFO(buffer),
                MessageIds.SID_AUTH_INFO => new SID_AUTH_INFO(buffer),
                MessageIds.SID_AUTH_CHECK => new SID_AUTH_CHECK(buffer),
                MessageIds.SID_SETEMAIL => new SID_SETEMAIL(buffer),
                MessageIds.SID_FRIENDSLIST => new SID_FRIENDSLIST(buffer),
                MessageIds.SID_FRIENDSUPDATE => new SID_FRIENDSUPDATE(buffer),
                MessageIds.SID_FRIENDSADD => new SID_FRIENDSADD(buffer),
                MessageIds.SID_FRIENDSREMOVE => new SID_FRIENDSREMOVE(buffer),
                MessageIds.SID_FRIENDSPOSITION => new SID_FRIENDSPOSITION(buffer),
                MessageIds.SID_CLANINFO => new SID_CLANINFO(buffer),
                MessageIds.SID_CLANMEMBERLIST => new SID_CLANMEMBERLIST(buffer),
                MessageIds.SID_CLANMEMBERSTATUSCHANGE => new SID_CLANMEMBERSTATUSCHANGE(buffer),
                /*
                MessageIds.SID_ => new SID_(buffer),
                */
                _ => null,
            };
        }

        public static string MessageName(byte messageId)
        {
            return ((MessageIds)messageId).ToString();
        }

        public byte[] ToByteArray(ProtocolType protocolType)
        {
            if (protocolType.IsGame())
            {
                var size = (UInt16)(4 + Buffer.Length);
                var buffer = new byte[size];

                buffer[0] = 0xFF;
                buffer[1] = Id;
                buffer[2] = (byte)(size);
                buffer[3] = (byte)(size >> 8);

                System.Buffer.BlockCopy(Buffer, 0, buffer, 4, Buffer.Length);

                return buffer;
            }
            else if (protocolType.IsChat())
            {
                return Encoding.UTF8.GetBytes($"{2000 + Id} {MessageName(Id).Replace("SID_", "")}{Battlenet.Common.NewLine}");
            }
            else
            {
                throw new NotSupportedException();
            }
        }

        public abstract bool Invoke(MessageContext context);
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Other/Message.vb
Imports AtlasV.Battlenet.Protocols.Game.Messages
Imports AtlasV.Battlenet.Exceptions
Imports System
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game
    MustInherit Class Message
        Public Id As Byte
        Public Property Buffer As Byte()

        Public Shared Function FromByteArray(ByVal varBuffer As Byte()) As Message
            If varBuffer(0) <> &HFF Then Throw New GameProtocolViolationException(Nothing, "Invalid message header")
            Dim id As Byte = varBuffer(1)
            Dim length As UInt16 = varBuffer(3) : length <<= 8 : length += varBuffer(2)
            Dim body As Byte() = New Byte(length - 4 - 1) {}
            System.Buffer.BlockCopy(varBuffer, 4, body, 0, length - 4)
            Return FromByteArray(id, body)
        End Function

        Public Shared Function FromByteArray(ByVal id As MessageIds, ByVal buffer As Byte()) As Message
            Select Case id
                Case MessageIds.SID_NULL : Return New SID_NULL(buffer)
                Case MessageIds.SID_STOPADV : Return New SID_STOPADV(buffer)
                Case MessageIds.SID_CLIENTID : Return New SID_CLIENTID(buffer)
                Case MessageIds.SID_STARTVERSIONING : Return New SID_STARTVERSIONING(buffer)
                Case MessageIds.SID_REPORTVERSION : Return New SID_REPORTVERSION(buffer)
                Case MessageIds.SID_STARTADVEX : Return New SID_STARTADVEX(buffer)
                Case MessageIds.SID_GETADVLISTEX : Return New SID_GETADVLISTEX(buffer)
                Case MessageIds.SID_ENTERCHAT : Return New SID_ENTERCHAT(buffer)
                Case MessageIds.SID_GETCHANNELLIST : Return New SID_GETCHANNELLIST(buffer)
                Case MessageIds.SID_JOINCHANNEL : Return New SID_JOINCHANNEL(buffer)
                Case MessageIds.SID_CHATCOMMAND : Return New SID_CHATCOMMAND(buffer)
                Case MessageIds.SID_CHATEVENT : Return New SID_CHATEVENT(buffer)
                Case MessageIds.SID_LEAVECHAT : Return New SID_LEAVECHAT(buffer)
                Case MessageIds.SID_LOCALEINFO : Return New SID_LOCALEINFO(buffer)
                Case MessageIds.SID_FLOODDETECTED : Return New SID_FLOODDETECTED(buffer)
                Case MessageIds.SID_UDPPINGRESPONSE : Return New SID_UDPPINGRESPONSE(buffer)
                Case MessageIds.SID_CHECKAD : Return New SID_CHECKAD(buffer)
                Case MessageIds.SID_CLICKAD : Return New SID_CLICKAD(buffer)
                Case MessageIds.SID_READMEMORY : Return New SID_READMEMORY(buffer) 'Lol guy
                Case MessageIds.SID_REGISTRY : Return New SID_REGISTRY(buffer)
                Case MessageIds.SID_MESSAGEBOX : Return New SID_MESSAGEBOX(buffer)
                Case MessageIds.SID_STARTADVEX2 : Return New SID_STARTADVEX2(buffer)
                Case MessageIds.SID_GAMEDATAADDRESS : Return New SID_GAMEDATAADDRESS(buffer)
                Case MessageIds.SID_STARTADVEX3 : Return New SID_STARTADVEX3(buffer)
                Case MessageIds.SID_LOGONCHALLENGEEX : Return New SID_LOGONCHALLENGEEX(buffer)
                Case MessageIds.SID_CLIENTID2 : Return New SID_CLIENTID2(buffer)
                Case MessageIds.SID_DISPLAYAD : Return New SID_DISPLAYAD(buffer)
                Case MessageIds.SID_NOTIFYJOIN : Return New SID_NOTIFYJOIN(buffer)
                Case MessageIds.SID_PING : Return New SID_PING(buffer)
                Case MessageIds.SID_READUSERDATA : Return New SID_READUSERDATA(buffer)
                Case MessageIds.SID_WRITEUSERDATA : Return New SID_WRITEUSERDATA(buffer)
                Case MessageIds.SID_LOGONCHALLENGE : Return New SID_LOGONCHALLENGE(buffer)
                Case MessageIds.SID_LOGONRESPONSE : Return New SID_LOGONRESPONSE(buffer)
                Case MessageIds.SID_CREATEACCOUNT : Return New SID_CREATEACCOUNT(buffer)
                Case MessageIds.SID_SYSTEMINFO : Return New SID_SYSTEMINFO(buffer)
                Case MessageIds.SID_GAMERESULT : Return New SID_GAMERESULT(buffer)
                Case MessageIds.SID_GETICONDATA : Return New SID_GETICONDATA(buffer)
                Case MessageIds.SID_CDKEY : Return New SID_CDKEY(buffer)
                Case MessageIds.SID_CHECKDATAFILE : Return New SID_CHECKDATAFILE(buffer)
                Case MessageIds.SID_GETFILETIME : Return New SID_GETFILETIME(buffer)
                Case MessageIds.SID_CDKEY2 : Return New SID_CDKEY2(buffer)
                Case MessageIds.SID_LOGONRESPONSE2 : Return New SID_LOGONRESPONSE2(buffer)
                Case MessageIds.SID_CHECKDATAFILE2 : Return New SID_CHECKDATAFILE2(buffer)
                Case MessageIds.SID_CREATEACCOUNT2 : Return New SID_CREATEACCOUNT2(buffer)
                Case MessageIds.SID_QUERYADURL : Return New SID_QUERYADURL(buffer)
                Case MessageIds.SID_WARCRAFTGENERAL : Return New SID_WARCRAFTGENERAL(buffer)
                Case MessageIds.SID_NEWS_INFO : Return New SID_NEWS_INFO(buffer)
                Case MessageIds.SID_AUTH_INFO : Return New SID_AUTH_INFO(buffer)
                Case MessageIds.SID_AUTH_CHECK : Return New SID_AUTH_CHECK(buffer)
                Case MessageIds.SID_FRIENDSLIST : Return New SID_FRIENDSLIST(buffer)
                Case MessageIds.SID_FRIENDSUPDATE : Return New SID_FRIENDSUPDATE(buffer)
                Case MessageIds.SID_FRIENDSADD : Return New SID_FRIENDSADD(buffer)
                Case MessageIds.SID_FRIENDSREMOVE : Return New SID_FRIENDSREMOVE(buffer)
                Case MessageIds.SID_FRIENDSPOSITION : Return New SID_FRIENDSPOSITION(buffer)
                Case MessageIds.SID_CLANINFO : Return New SID_CLANINFO(buffer)
                Case MessageIds.SID_CLANMEMBERLIST : Return New SID_CLANMEMBERLIST(buffer)
                Case MessageIds.SID_CLANMEMBERSTATUSCHANGE : Return New SID_CLANMEMBERSTATUSCHANGE(buffer)
                Case Else : Return Nothing
            End Select
        End Function

        Public Shared Function MessageName(varMessageId As MessageIds) As String
            Return (varMessageId).ToString()
        End Function

        Public Function ToByteArray(ByVal varProtocolType As ProtocolType) As Byte()
            If varProtocolType.IsGame() Then
                Dim size = CUShort((4 + Buffer.Length)) 'you fuckin need to properly name your vars wtf
                Dim rereBuffer = New Byte(size - 1) {}
                rereBuffer(0) = &HFF
                rereBuffer(1) = Id
                rereBuffer(2) = CByte((size))
                rereBuffer(3) = CByte(size >> 8)
                System.Buffer.BlockCopy(Buffer, 0, rereBuffer, 4, Buffer.Length)
                Return rereBuffer
            ElseIf varProtocolType.IsChat() Then
                Return Encoding.UTF8.GetBytes($"{2000 + Id} {MessageName(Id).Replace("SID_", "")}{Battlenet.Common.NewLine}")
            Else
                Throw New NotSupportedException()
            End If
        End Function

        Public MustOverride Function Invoke(ByVal context As MessageContext) As Boolean

    End Class
End Namespace